lib/redlock/client.rb

```diff
@@ -228,6 +281,37 @@ def lock_instances(resource, ttl, options)
       end
     end
 
+    def try_get_remaining_ttl(resource)
+      # Responses from the servers are a 2 tuple of format [lock_value, ttl].
+      # The lock_value is nil if it does not exist. Since servers may have
+      # different lock values, the responses are grouped by the lock_value and
+      # transofrmed into a hash: { lock_value1 => [ttl1, ttl2, ttl3],
+      # lock_value2 => [ttl4, tt5] }
+      ttls_by_value, time_elapsed = timed do
+        @servers.map { |s| s.get_remaining_ttl(resource) }
+          .select(&:first)
+          .group_by(&:first)
+          .transform_values { |ttl_tuples| ttl_tuples.map { |t| t[1] } }
+      end
+
+      # Authoritative lock value is that which is returned by the majority of
+      # servers
+      authoritative_value, ttls =
+        ttls_by_value.max_by { |(lock_value, ttls)| ttls.length }
+
+      if ttls && ttls.size >= @quorum
+        # Return the  minimum TTL of an N/2+1 selection. It will always be
+        # correct (it will guarantee that at least N/2+1 servers have a TTL that
+        # value or longer)
+        min_ttl = ttls.sort.last(@quorum).min
```

Comment:
technically this is the same as `ttls.sort.last(@quorum).first`, isn't it? saves you one list scan to get the `min`